activerecord/lib/active_record/attribute_methods/composite_primary_key.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  module AttributeMethods
+    module CompositePrimaryKey # :nodoc:
+      # Returns the primary key column's value. If the primary key is composite,
+      # returns an array of the primary key column values.
+      def id
+        if self.class.composite_primary_key?
```

Comment:
I benched this vs `@primary_key.is_a?(Array)`, the difference is fairly minimal (5-7%):

```
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [arm64-darwin22]
Warming up --------------------------------------
      CPK#id (is_a?)   200.372k i/100ms
      CPK#id (class)   190.537k i/100ms
Calculating -------------------------------------
      CPK#id (is_a?)      1.993M (± 2.6%) i/s -     10.019M in   5.029710s
      CPK#id (class)      1.855M (± 7.8%) i/s -      9.336M in   5.075258s

Comparison:
      CPK#id (is_a?):  1993425.4 i/s
      CPK#id (class):  1854978.2 i/s - same-ish: difference falls within error
```

But given that with CPK `#id` is already more than twice slower than without, I don't think it's worth micro-optimizing it.

If we really want to optimize it, one solution would be to disallow setting a non-CPK on a child of a model with CPK, as it make really little sense to I wouldn't mind having such limitation. It would be a bit tricky to detect as there is potentially ordering issues.

But even that doesn't seem to make that much of a difference:

```
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [arm64-darwin22]
Warming up --------------------------------------
   CPK#id (no-check)   208.409k i/100ms
      CPK#id (class)   186.005k i/100ms
Calculating -------------------------------------
   CPK#id (no-check)      2.040M (± 5.8%) i/s -     10.212M in   5.027149s
      CPK#id (class)      1.882M (± 3.3%) i/s -      9.486M in   5.047891s

Comparison:
   CPK#id (no-check):  2039769.0 i/s
      CPK#id (class):  1881645.3 i/s - same-ish: difference falls within error
```

```ruby
begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "rails", path: "/Users/byroot/src/github.com/Shopify/rails"
  gem "sqlite3"
  gem "benchmark-ips"
end

require "active_record"

ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :users do |t|
    t.string :first_name
    t.string :last_name
  end
end

class User < ActiveRecord::Base
end

class CPKUser < User
  self.primary_key = [:id, :first_name]
end

module ActiveRecord
  module AttributeMethods
    module CompositePrimaryKey # :nodoc:
      # Returns the primary key column's value. If the primary key is composite,
      # returns an array of the primary key column values.
      def no_check_id
        @primary_key.map { |pk| _read_attribute(pk) }
      end

      def check_class_id
        if self.class.composite_primary_key?
          @primary_key.map { |pk| _read_attribute(pk) }
        else
          super
        end
      end
    end
  end
end

user = User.new
cpk_user = CPKUser.new

puts RUBY_DESCRIPTION

Benchmark.ips do |x|
  x.report("CPK#id (no-check)") do
    cpk_user.no_check_id
  end

  x.report("CPK#id (class)") do
    cpk_user.check_class_id
  end

  x.compare!(order: :baseline)
end
```